File: STFloss08/STFloss08.R
# clear variables and close windows
rm(list = ls(all = TRUE))
graphics.off()

samplemef = function(data, xaxis) {
    # SAMPLEMEF Sample mean excess function.  RES = SAMPLEMEF (DATA, XAXIS) returns the value of the sample mean excess function
    # for the vector DATA in the points from vector XAXIS.
    
    
    sorteddata = sort(data)
    dataLength = length(sorteddata)
    resLength = length(xaxis)
    res = xaxis
    
    
    i = 0
    valuesOnTheLeft = 0
    smef = mean(sorteddata[1:dataLength])
    
    while (i < resLength) {
        i = i + 1
        while (sorteddata[valuesOnTheLeft + 1] < xaxis[i]) {
            valuesOnTheLeft = valuesOnTheLeft + 1
        }
        smef = mean(sorteddata[(valuesOnTheLeft + 1):dataLength])
        res[i] = smef - xaxis[i]
    }
    return(res)
}


# load the data
data = read.table("dfl.txt")
data = data[, 5]
xaxis = sort(data)

# calculate empirical mean excess function on dfl data
sampleMef = samplemef(data, xaxis)
sampleMef = sampleMef/1e+06

xaxis = xaxis/1e+06

plot(xaxis, sampleMef, col = "black", xlab = "x (DKK million)", ylab = "e_n(x) (DKK million)", xlim = c(0, 18), ylim = c(0, 
    max(sampleMef))) 

File: STFloss08/STFloss08.m

% clear variables and close windows
clear all
close all
clc


% load the data
data = load('dfl.dat','-ascii');
data = data(:,5);
xaxis = sort(data);

sampleMef = samplemef (data, xaxis);
sampleMef = sampleMef / 1e6;

xaxis = xaxis / 1e6;
figure
plot(xaxis,sampleMef,'ok');
ylabel('e_n(x) (DKK million)','FontSize',16,'FontWeight','Bold');
xlabel('x (DKK million)','FontSize',16,'FontWeight','Bold');
xlim([0,18]);
set(gca,'LineWidth',1.6,'FontSize',16,'FontWeight','Bold');
set(gca,'Ytick',[0:2:12],'YTickLabel',{0,2,4,6,8,10,12},'FontSize',16,'FontWeight','Bold')
box on

% to save the plot in pdf or png please uncomment next 2 lines:
%print -painters -dpdf -r600 STFloss08.pdf
%print -painters -dpng -r600 STFloss08.png
